src/Struct.php

```diff
@@ -0,0 +1,223 @@
+ *
+ * Lastly, this class only reserves one property name: "__data". The property is prefixed with a double underscore to
+ * reduce the probability of conflicting with any consumer property name. However, it is recommended to avoid to using
+ * double underscore prefixes as additional properties may be added in future versions to accommodate new features.
+ *
+ * @since [*next-version*]
+ */
+abstract class Struct implements Serializable
+{
+    /**
+     * @since [*next-version*]
+     *
+     * @var array
+     */
+    protected $__data;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param array $data A mapping of property names to their values.
+     */
+    public function __construct(array $data = [])
+    {
+        $props = $this->getPropTypes();
+
+        $this->__data = [];
+        foreach ($props as $key => $type) {
+            $this->__data[$key] = isset($data[$key])
+                ? $type->cast($data[$key])
+                : $type->getDefault();
+
+            unset($data[$key]);
+        }
+
+        if (!empty($data)) {
+            $class = static::class;
+            $props = implode(', ', array_keys($data));
+            throw new LogicException("Struct {$class} does not have the following properties: {$props}");
+        }
+    }
+
+    /**
+     * Retrieves the value for a property.
+     *
+     * @since [*next-version*]
+     *
+     * @param string $name The name of the property.
+     *
+     * @return mixed|null The value of the property.
+     */
+    public function __get($name)
+    {
+        $props = $this->getPropTypes();
+
+        if (!array_key_exists($name, $props)) {
+            throw static::createUndefinedPropException($name);
+        }
+
+        return $this->__data[$name];
+    }
+
+    /**
+     * Should not be used. Only exists to warn developers of misuse.
+     *
+     * @since [*next-version*]
+     *
+     * @param string $name  The name of the property.
+     * @param mixed  $value The value being assigned to the property.
+     *
+     * @throws Exception Always throws.
+     */
+    public function __set($name, $value)
+    {
+        throw new LogicException('Cannot set property values. Use with() instead');
+    }
+
+    /**
+     * Creates a copy of the struct instance with some modified properties.
+     *
+     * @since [*next-version*]
+     *
+     * @param array $changes An associative array that maps property names to their new values.
+     *
+     * @return Struct The copied struct with the applied changes. If the $changes parameter is empty, no copy is
+     *                performed and the same instance is returned.
+     */
+    public function with(array $changes)
+    {
+        if (empty($changes)) {
+            return $this;
+        }
+
+        $clone = clone $this;
+
+        $props = $this->getPropTypes();
+        foreach ($changes as $key => $value) {
+            if (!isset($props[$key])) {
```

Comment:
The kind of stuff that happens here seems to be similar to what's happening in the constructor. I wouldn't recommend creating a new instance of `$this`, instead of `clone`, because that may miss some non-injectable data. But what about adding a `protected` method that does this, and invoke it here and in the constructor?

Reply:
[See this commit](https://github.com/Dhii/structs/pull/1/commits/c1b7e17b2436bd436fb3b66126f70786c3320eb9).

Reply:
Re. creating a new instance of `$this`, I tried avoiding being bound to the constructor to let descendants of `Struct` implement their own constructor if they require it.

Reply:
Some additional explanation to add onto what I said in the commit message:

The ctor's job is to set the internal data set. That involves iterating the struct's prop types and for each prop type retrieve the corresponding value from the ctor argument, casting it if given and using the default value if not.

The `with()` method however does something different. It iterates over the argument, not the internal prop types, and casts each value using the corresponding prop type.

The algorithms _are_ similar, true, but fundamentally different. It _is_ possible to create a centralized algorithm, but it would make either the ctor or the `with()` method inefficient, depending on what is iterated over. I wasn't willing to make such a compromise.